extensions/amp-truncate-text/0.1/amp-truncate-text.js

```diff
@@ -0,0 +1,259 @@
+
+  /** @override */
+  buildCallback() {
+    this.useShadow_ = !!this.element.attachShadow &&
+      isExperimentOn(this.win, 'amp-truncate-text-shadow');
+
+    if (this.useShadow_) {
+      this.buildShadow_();
+    } else {
+      this.build_();
+    }
+
+    if ('MutationObserver' in window) {
+      this.mutationObserver_ = new MutationObserver(() => {
+        this.truncate_();
+      });
+    }
+
+    this.expandSlot_.addEventListener('click', () => this.expand_());
+    this.collapseSlot_.addEventListener('click', () => this.collapse_());
+  }
+
+  /**
+   * Builds the component when not using Shadow DOM.
+   */
+  build_() {
+    const html = htmlFor(this.element);
+    this.content_ = html`
+      <div class="i-amphtml-truncate-content">
+        <span class="i-amphtml-default-slot"></span>
+        <span class="i-amphtml-truncate-expand-slot" name="expand"></span>
+        <span class="i-amphtml-truncate-collapse-slot" name="collapse"></span>
+      </div>
+    `;
+
+    const defaultSlot = this.content_.querySelector('.i-amphtml-default-slot');
+    this.expandSlot_ = this.content_
+        .querySelector('.i-amphtml-truncate-expand-slot');
+    this.collapseSlot_ = this.content_
+        .querySelector('.i-amphtml-truncate-collapse-slot');
+
+    iterateCursor(this.element.querySelectorAll('[slot="expand"]'), el => {
+      this.expandSlot_.appendChild(el);
+    });
+    iterateCursor(this.element.querySelectorAll('[slot="collapse"]'), el => {
+      this.collapseSlot_.appendChild(el);
+    });
+    this.getRealChildNodes().forEach(node => {
+      defaultSlot.appendChild(node);
+    });
+
+    this.element.appendChild(this.content_);
+  }
+
+  /**
+   * Builds the component when using Shadow DOM.
+   */
+  buildShadow_() {
+    // TODO(sparhami) Where is the right place to put this? Runtime? What about
+    // SSR?
+    const sizer = this.element.querySelector('i-amphtml-sizer');
+    if (sizer) {
+      sizer.setAttribute('slot', 'sizer');
+    }
+
+    // TODO(sparhami) Is there a shared place to add logic for creating
+    // shadow roots with styles? Might make sense to have it create the style
+    // as well as a slot for the sizer.
+    const sr = this.element.attachShadow({mode: 'open'});
+    const style = document.createElement('style');
+    style.textContent = ShadowCSS;
+    const html = htmlFor(this.element);
+    const content = html`
+      <div>
+        <div class="content">
+          <slot></slot>
+          <slot class="expand-slot" name="expand"></slot>
+          <slot class="collapse-slot" name="collapse"></slot>
+        </div>
+        <slot name="sizer"></slot>
+      </div>
+    `;
+
+    sr.appendChild(style);
+    sr.appendChild(content);
+
+    this.content_ = null;
+    this.expandSlot_ = content.querySelector('.expand-slot');
+    this.collapseSlot_ = content.querySelector('.collapse-slot');
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (isExperimentOn(this.win, 'amp-truncate-text')) {
```

Comment:
Seems like most of work is happening in the `buildCallback`. We should put experiment check and throw an exception there. Otherwise it could be easy not to notice...